docs: Add more documentation (#77)
keptn-sandbox/lifecycle-controller: README.md

# Keptn Lifecycle Controller

![build](https://img.shields.io/github/workflow/status/keptn-sandbox/lifecycle-controller/CI)
![goversion](https://img.shields.io/github/go-mod/go-version/keptn-sandbox/lifecycle-controller?filename=operator%2Fgo.mod)
![version](https://img.shields.io/badge/version-pre--alpha-green)
![status](https://img.shields.io/badge/status-not--for--production-red)
[![GitHub Discussions](https://img.shields.io/github/discussions/keptn-sandbox/lifecycle-controller)](https://github.com/keptn-sandbox/lifecycle-controller/discussions)

The purpose of this repository is to demonstrate and experiment with
a prototype of a _**Keptn Lifecycle Controller**_.
The goal of this prototype is to introduce a more “cloud-native” approach for pre- and post-deployment, as well as the concept of application health checks.
It is an experimental project, under the umbrella of the [Keptn Application Lifecycle working group](https://github.com/keptn/wg-app-lifecycle).

## Deploy the latest release

> The lifecycle controllers includes a Mutating Webhook which requires TLS certificates to be mounted as a volume in its pod. The certificate creation

Reviewer: ```suggestion
The lifecycle controller includes a Mutating Webhook which requires TLS certificates to be mounted as a volume in its pod. The certificate creation
```

is handled automatically by [cert-manager](https://cert-manager.io). To install **cert-manager**, follow their [installation instructions](https://cert-manager.io/docs/installation/).

> When the *cert-manager* is installed, you can run

Reviewer: ```suggestion
When *cert-manager* is installed, you can run
```

<!---x-release-please-start-version-->

```
kubectl apply -f https://github.com/keptn-sandbox/lifecycle-controller/releases/download/v0.1.0/release.yaml
```

<!---x-release-please-end-->

to install the latest release of the lifecycle controller.

## Goals

> The Keptn Lifecycle Controller aims at supporting Cloud Native teams with:

Reviewer: ```suggestion
The Keptn Lifecycle Controller aims to support Cloud Native teams with:
```

> - Pre-requisite evaluation before deploying Workloads and applications

Reviewer: ```suggestion
- Pre-requisite evaluation before deploying workloads and applications
```

- Finding out when an application (not workload) is ready and working
- Checking the Application Health in a declarative (cloud-native) way
- Standardized way for pre- and post-deployment tasks
- Provide out-of-the-box Observability of the deployment cycle

![](./assets/operator-maturity.jpg)

> The Keptn Lifecycle Controller could be seen as a general purpose and declarative Level 3 operator for your Application.

Reviewer: Maybe provide a link to operator maturity docs to explain the levels?

> For this reason, the Keptn Lifecycle Controller is agnostic to the Deployment tool that is being used and works with any GitOps solution.

Reviewer: ```suggestion
For this reason, the Keptn Lifecycle Controller is agnostic to deployment tools that are used and works with any GitOps solution.
```

## How to use

> The Keptn Lifecycle Controller monitors manifests that have been applied against the KubeAPI and reacts if it finds a workload with special annotations.

Reviewer: ```suggestion
The Keptn Lifecycle Controller monitors manifests that have been applied against the Kubernetes API and reacts if it finds a workload with special annotations.
```

For this, you should annotate your [Workload](https://kubernetes.io/docs/concepts/workloads/) with (at least) the following two annotations:

```yaml
keptn.sh/app: myAwesomeAppName
keptn.sh/workload: myAwesomeWorkload
```

> In the case you want to run pre- and post-deployment checks, further annotations are necessary:

Reviewer: ```suggestion
In case you want to run pre- and post-deployment checks, further annotations are necessary:
```

```yaml
keptn.sh/pre-deployment-tasks: verify-infrastructure-problems
keptn.sh/post-deployment-tasks: slack-notification,performance-test
```

> In this example, before the deployment starts, a check for open problems in your infrastructure are performed. If everything is fine, the deployment continues and afterward, a slack notification is sent with the result of the deployment and a pipeline to run performance tests is invoked.

Reviewer 2: Is there a function for that or is this only an example? If this is only an example, we should write this there ...

> In this example, before the deployment starts, a check for open problems in your infrastructure are performed. If everything is fine, the deployment continues and afterward, a slack notification is sent with the result of the deployment and a pipeline to run performance tests is invoked.

Reviewer: ```suggestion
In this example, before the deployment starts, a check for open problems in your infrastructure is performed. If everything is fine, the deployment continues and afterwards, a slack notification is sent with the result of the deployment and a pipeline to run performance tests is invoked.
```

A more comprehensive example can be found in our [examples folder](./examples/podtatohead-deployment/) where we use [Podtato-Head](https://github.com/podtato-head/podtato-head) to run pre-deployment checks.

To run the example, use the following commands:

```bash
cd ./examples/podtatohead-deployment/
kubectl apply -f .
```

> Afterward, you can monitor the status of the deployment using

Reviewer: ```suggestion
Afterwards, you can monitor the status of the deployment using
```

Author: Let's keep the more known US version

The Keptn Scheduler ensures that Pods are started only after the pre-deployment checks have finished.

## Goals

The Keptn Lifecycle Controller aims at supporting Cloud Native teams with:

- Pre-requisite evaluation before deploying Workloads and applications
- Finding out when an application (not workload) is ready and working
- Checking the Application Health in a declarative (cloud-native) way
- Standardized way for pre- and post-deployment tasks
- Provide out-of-the-box Observability of the deployment cycle

![](./assets/operator-maturity.jpg)

The Keptn Lifecycle Controller could be seen as a general purpose and declarative Level 3 operator for your Application.
For this reason, the Keptn Lifecycle Controller is agnostic to the Deployment tool that is being used and works with any GitOps solution.

## How to install

**Prerequisites:**

The lifecycle controllers includes a Mutating Webhook which requires TLS certificates to be mounted as a volume in its pod. The certificate creation
is handled automatically by [cert-manager](https://cert-manager.io). To install **cert-manager**, follow their [installation instructions](https://cert-manager.io/docs/installation/).

> When the *cert-manager* is installed, you can run

Reviewer 3: If you put a header for prerequisites, another one should be here

```bash
kubectl apply -f <insert link of release>
```

## How to use

> The Keptn Lifecycle Controller monitors manifests that are applied against the KubeAPI and reacts if it finds a workload with special annotations.

Reviewer 2: That's not entirely correct, as manifests are already applied when pods are changed by the mutating webhook ...

Next, it will look for an existing instance of a `Workload CRD` for the given workload name:

- If it finds the `Workload`, it will update its version according to the previously computed version string. In addition, it will include a reference to the ReplicaSet UID of the pod (i.e. the Pods owner), or the pod itself, if it does not have an owner.

> - If it does not find a workload instance, it will create one containing the previously computed version string. In addition, it will include a reference to the ReplicaSet UID of the pod (i.e. the Pods owner), or the pod itself, if it does not have an owner.

Reviewer 3: Protip: split paragraph and sentences to multiple lines to make it more maintenable

### Scheduler

> After the Webhook mutation, the Keptn-Scheduler will handle the annotated resources. The scheduling flow follows the default scheduler,

Reviewer: ```suggestion
After the Webhook mutation, the Keptn-Scheduler will handle the annotated resources. The scheduling flow follows the default scheduler behavior,
```

Workload Instances have a reference to the respective Deployment/StatefulSet/ReplicaSet, to check if it has reached the desired state. If it detects that the referenced object has reached
its desired state (e.g. all pods of a deployment are up and running), it will be able to tell that a `PostDeploymentCheck` can be triggered.

### Keptn Task Definition

A `KeptnTaskDefinition` is a CRD used to define tasks that can be run by the Keptn Lifecycle Controller
as part of pre- and post-deployment phases of a deployment.
The task definition is a [Deno](https://deno.land/) script
Please, refer to the [function runtime](./functions-runtime/) folder for more information about the runtime.

> In the future, we will support also other runtime, especially running a container image directly.

Reviewer 3: We intend to.
I am not sure it is a good idea for all cases

A task definition can be configured in three different way:

- inline
- referring an HTTP script
- referring another `KeptnTaskDefinition`

An inline task definition looks like the following:

```yaml
apiVersion: lifecycle.keptn.sh/v1alpha1
kind: KeptnTaskDefinition
metadata:
  name: deployment-hello
spec:
  function:
    inline:
      code: |
        console.log("Deployment Task has been executed");
```

> In the code section, it is possible to define a full-fletch Deno script.

Reviewer 3: Full-fletched

A further example, is available [here](./examples/taskonly-hello-keptn/inline/taskdefinition.yaml).

> To runtime can also fetch the script on the fly from a remote webserver. For this, the CRD should look like the following:

Reviewer 3: Needs explicit security considerations in the documentation. Can be done in a follow up pull request

```yaml
apiVersion: lifecycle.keptn.sh/v1alpha1
kind: KeptnTaskDefinition
metadata:
  name: hello-keptn-http
spec:
  function:
    httpRef:
      url: <url>
```

An example is available [here](./examples/taskonly-hello-keptn/http/taskdefinition.yaml).

Finally, `KeptnTaskDefinition` can build on top of other `KeptnTaskDefinition`.
This is a common use case where a general function can be re-used in multiple places with different parameters.

```yaml
apiVersion: lifecycle.keptn.sh/v1alpha1
kind: KeptnTaskDefinition
metadata:
  name: slack-notification-dev
spec:
  function:
    functionRef:
      name: slack-notification
    parameters:
      map:
        textMessage: "This is my configuration"
    secureParameters:
      secret: slack-token
```

As you might have noticed, Task Definitions have also the possibility to have parameters.

> The Lifecycle Controller passes the values defined inside the `map` field a JSON object.

Reviewer 3: Is multi-level map allowed (when you put nested objects to the keys and objects)

Author: I added the limitation that we don't support them

Workload Instances have a reference to the respective Deployment/StatefulSet/ReplicaSet, to check if it has reached the desired state. If it detects that the referenced object has reached
its desired state (e.g. all pods of a deployment are up and running), it will be able to tell that a `PostDeploymentCheck` can be triggered.

### Keptn Task Definition

A `KeptnTaskDefinition` is a CRD used to define tasks that can be run by the Keptn Lifecycle Controller
as part of pre- and post-deployment phases of a deployment.
The task definition is a [Deno](https://deno.land/) script

> Please, refer to the [function runtime](./functions-runtime/) folder for more information about the runtime.

Reviewer: ```suggestion
Please refer to the [function runtime](./functions-runtime/) folder for more information about the runtime.
```

> In the future, we intend to support also other runtime, especially running a container image directly.

Reviewer: ```suggestion
In the future, we also intend to support other runtimes, especially running a container image directly.
```

> A task definition can be configured in three different way:
> 
> - inline
> - referring an HTTP script
> - referring another `KeptnTaskDefinition`

Reviewer: ```suggestion
A task definition can be configured in three different ways:

- inline
- referring to an HTTP script
- referring to another `KeptnTaskDefinition`
```

An inline task definition looks like the following:

```yaml
apiVersion: lifecycle.keptn.sh/v1alpha1
kind: KeptnTaskDefinition
metadata:
  name: deployment-hello
spec:
  function:
    inline:
      code: |
        console.log("Deployment Task has been executed");
```

In the code section, it is possible to define a full-fletched Deno script.
A further example, is available [here](./examples/taskonly-hello-keptn/inline/taskdefinition.yaml).

To runtime can also fetch the script on the fly from a remote webserver. For this, the CRD should look like the following:

```yaml
apiVersion: lifecycle.keptn.sh/v1alpha1
kind: KeptnTaskDefinition
metadata:
  name: hello-keptn-http
spec:
  function:
    httpRef:
      url: <url>
```

An example is available [here](./examples/taskonly-hello-keptn/http/taskdefinition.yaml).

> Finally, `KeptnTaskDefinition` can build on top of other `KeptnTaskDefinition`.

Reviewer: ```suggestion
Finally, `KeptnTaskDefinition` can build on top of other `KeptnTaskDefinition`s.
```

This is a common use case where a general function can be re-used in multiple places with different parameters.

```yaml
apiVersion: lifecycle.keptn.sh/v1alpha1
kind: KeptnTaskDefinition
metadata:
  name: slack-notification-dev
spec:
  function:
    functionRef:
      name: slack-notification
    parameters:
      map:
        textMessage: "This is my configuration"
    secureParameters:
      secret: slack-token
```

> As you might have noticed, Task Definitions have also the possibility to have parameters.

Reviewer: ```suggestion
As you might have noticed, Task Definitions also have the possibility to use input parameters.
```

> The Lifecycle Controller passes the values defined inside the `map` field a JSON object.

Reviewer: ```suggestion
The Lifecycle Controller passes the values defined inside the `map` field as a JSON object.
```

At the moment, multi-level maps are not supported.

> The JSON object can be read through the env var `DATA` using `Deno.env.get("DATA");`.

Reviewer: ```suggestion
The JSON object can be read through the environment variable `DATA` using `Deno.env.get("DATA");`.
```

K8s secrets can also be passed to the function using the `secureParameters` field.

> Here, the `secret` value is the K8s secret name that will be mounted into the runtime and made available to the function via the env var `SECURE_DATA`.

Reviewer: ```suggestion
Here, the `secret` value is the K8s secret name that will be mounted into the runtime and made available to the function via the environment variable `SECURE_DATA`.
```

### Keptn Task

A Task is responsible for executing the TaskDefinition of a workload.
The execution is done spawning a K8s Job to handle a single Task.
In its state, it keeps track of the current status of the K8s Job created.

## Install a dev build

> The [GH CLI](https://cli.github.com/) can be used to download the manifests of the latest CI build.

Reviewer: ```suggestion
The [GitHub CLI](https://cli.github.com/) can be used to download the manifests of the latest CI build.
```
